velox/expression/CastExpr.cpp

```diff
@@ -147,18 +148,39 @@ VectorPtr castToDate(
     case TypeKind::TIMESTAMP: {
       auto* inputVector = input.as<SimpleVector<Timestamp>>();
       static const int32_t kSecsPerDay{86'400};
-      context.applyToSelectedNoThrow(rows, [&](int row) {
-        auto input = inputVector->valueAt(row);
-        auto seconds = input.getSeconds();
-        if (seconds >= 0 || seconds % kSecsPerDay == 0) {
-          resultFlatVector->set(row, seconds / kSecsPerDay);
-        } else {
-          // For division with negatives, minus 1 to compensate the discarded
-          // fractional part. e.g. -1/86'400 yields 0, yet it should be
-          // considered as -1 day.
-          resultFlatVector->set(row, seconds / kSecsPerDay - 1);
-        }
-      });
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      auto sessionTzName = queryConfig.sessionTimezone();
+      if (queryConfig.adjustTimestampToTimezone() && !sessionTzName.empty()) {
+        // locate_zone throws runtime_error if the timezone couldn't be found
+        // (so we're safe to dereference the pointer).
+        auto* timeZone = date::locate_zone(sessionTzName);
+        context.applyToSelectedNoThrow(rows, [&](int row) {
```

Comment:
This and the loop below differ in only one statement "input.toTimezone(*timeZone)". Maybe abstract a common function for the loop with a boolean template parameter that can be used to control the timezone conversion.

Reply:
Thanks, will address this.